draft-ietf-masque-connect-udp-listen.md

```diff
@@ -154,10 +178,114 @@ UDP Payload:
 : The unmodified UDP Payload of this proxied UDP packet (referred to as "data
 octets" in {{UDP}}).
 
-# The Connect-UDP-Bind Header Field {#hdr}
+# Context ID {#contextid}
+
+This extension leverages context IDs (see {{Section 4 of CONNECT-UDP}}) to compress the target IP address and port when encoding datagrams on the wire. Either endpoint can register a context ID and the IP/ports it's associated with by sending a COMPRESSION_ASSIGN capsule to its peer. The peer will then echo that capsule to indicate it's received it. From then on, both endpoints are aware of the context ID and can send compressed datagrams. Later, any endpoint can decide to close the compression context by sending a COMPRESSION_CLOSE capsule.
+
+The context ID 0 was reserved by unextended connect-udp and is not used by this extension.
+Once an endpoint has ascertained that the peer supports this extension, the endpoint MUST NOT
+send any datagrams with context ID set to 0, and MUST drop any received datagrams with
+context ID set to 0.
+
+An even context ID MUST be used by clients to allocate uncompressed or compressed
+contexts, and an odd context ID MUST be used by the proxy to allocate compressed
+context IDs.They MAY pre-emptively use Context IDs not yet acknowledged by the other party, knowing that those packets MAY be lost since the COMPRESSION_ASSIGN request receiving proxy
+or client is not guaranteed to be ready to accept payloads until a COMPRESSION_ASSIGN
+response is echoed back.
+
+## Address Compression {#compression}
+
+The client and the proxy MAY choose to compress the IP and port information
+per datagram for a given target against the Context ID.
+In such a case, the client or the proxy MUST send a COMPRESSION_ASSIGN capsule
+(see {{capsuleassignformat}}) with the target information (see
+{{targetmappingformat}}) it wishes to compress and the other party (proxy or
+client respectively) MUST echo back with either a COMPRESSION_ASSIGN capsule
```

Comment:
This echoing MUST is redundant with the following section, can we refactor this to avoid duplicate normative language? For example we can have this paragraph explain the mechanism without normative text, and have the next section go into more details with 2119 language